pkg/reconciler/events/pubsub/pubsub.go

```diff
@@ -63,89 +60,99 @@ type Reconciler struct {
 // Check that we implement the controller.Reconciler interface.
 var _ controller.Reconciler = (*Reconciler)(nil)
 
-// Reconcile implements controller.Reconciler
+// Reconcile compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the PubSub resource
+// with the current status of the resource.
 func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
 	// Convert the namespace/name string into a distinct namespace and name
 	namespace, name, err := cache.SplitMetaNamespaceKey(key)
 	if err != nil {
-		runtime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		logging.FromContext(ctx).Desugar().Error("Invalid resource key")
 		return nil
 	}
 
 	// Get the PubSub resource with this namespace/name
 	original, err := r.pubsubLister.PubSubs(namespace).Get(name)
 	if apierrs.IsNotFound(err) {
 		// The PubSub resource may no longer exist, in which case we stop processing.
-		runtime.HandleError(fmt.Errorf("storage '%s' in work queue no longer exists", key))
+		logging.FromContext(ctx).Desugar().Error("PubSub in work queue no longer exists")
 		return nil
 	} else if err != nil {
 		return err
 	}
 
 	// Don't modify the informers copy
-	csr := original.DeepCopy()
+	pubsub := original.DeepCopy()
+
+	reconcileErr := r.reconcile(ctx, pubsub)
 
-	reconcileErr := r.reconcile(ctx, csr)
+	// If no error is returned, mark the observed generation.
+	if reconcileErr == nil {
+		pubsub.Status.ObservedGeneration = pubsub.Generation
+	}
 
-	if equality.Semantic.DeepEqual(original.Status, csr.Status) &&
-		equality.Semantic.DeepEqual(original.ObjectMeta, csr.ObjectMeta) {
+	if equality.Semantic.DeepEqual(original.Status, pubsub.Status) {
 		// If we didn't change anything then don't call updateStatus.
 		// This is important because the copy we loaded from the informer's
 		// cache may be stale and we don't want to overwrite a prior update
 		// to status with this stale state.
-	} else if _, err := r.updateStatus(ctx, csr); err != nil {
-		// TODO: record the event (c.Recorder.Eventf(...
-		r.Logger.Warn("Failed to update PubSub Source status", zap.Error(err))
-		return err
-	}
 
+	} else if _, uErr := r.updateStatus(ctx, pubsub); uErr != nil {
+		logging.FromContext(ctx).Desugar().Warn("Failed to update PubSub status", zap.Error(uErr))
+		r.Recorder.Eventf(pubsub, corev1.EventTypeWarning, "UpdateFailed",
+			"Failed to update status for PubSub %q: %v", pubsub.Name, uErr)
+		return uErr
+	} else if reconcileErr == nil {
+		// There was a difference and updateStatus did not return an error.
+		r.Recorder.Eventf(pubsub, corev1.EventTypeNormal, "Updated", "Updated PubSub %q", pubsub.Name)
+	}
 	if reconcileErr != nil {
-		// TODO: record the event (c.Recorder.Eventf(...
-		return reconcileErr
+		r.Recorder.Event(pubsub, corev1.EventTypeWarning, "InternalError", reconcileErr.Error())
```

Comment:
Should we also log the error in controller logs or is that done at a higher level? (and if it is done at a higher level, should we remove the log at L101?)

Reply:
Currently, the reconcileError is logged within the reconcile method. In this case, the functions called from within reconcile. I think I've done this all over the reconcilers.
The error in L101 is when there is a failure updating the status of the object. The update can cause an error on its own. It can be the case that there was a reconcileErr, and we also failed to update the status. We log both. 
I filed an issue to later on add retries for UpdateStatus. See https://github.com/google/knative-gcp/issues/442. I've seen it happen quite often actually.